src/combinatorics/simple_graph/matching.lean

```diff
@@ -1,69 +1,82 @@
-  - Hall's Marriage Theorem (see combinatorics.hall)
-  - Tutte's Theorem
-  - consider coercions instead of type definition for `matching`:
-    https://github.com/leanprover-community/mathlib/pull/5156#discussion_r532935457
-  - consider expressing `matching_verts` as union:
-    https://github.com/leanprover-community/mathlib/pull/5156#discussion_r532906131
+## Todo
 
-TODO: Tutte and Hall require a definition of subgraphs.
--/
-open finset
-universe u
+* Lemma stating that the existence of a perfect matching on `G` implies that
+  the cardinality of `V` is even (assuming it's finite)
 
-namespace simple_graph
-variables {V : Type u} (G : simple_graph V)
+* Hall's Marriage Theorem (see combinatorics.hall)
 
-/--
-A matching on `G` is a subset of its edges such that no two edges share a vertex.
+* Tutte's Theorem
+
+* https://github.com/leanprover-community/mathlib/pull/5156#discussion_r532906131
 -/
-structure matching :=
-(edges : set (sym2 V))
-(sub_edges : edges ⊆ G.edge_set)
-(disjoint : ∀ (x y ∈ edges) (v : V), v ∈ x ∧ v ∈ y → x = y)
 
-instance : inhabited (matching G) :=
-⟨⟨∅, set.empty_subset _, λ _ _ hx, false.elim (set.not_mem_empty _ hx)⟩⟩
+universe u
 
-variables {G}
+namespace simple_graph
+variables {V : Type u} {G : simple_graph V} (M : subgraph G)
 
 /--
-`M.support` is the set of vertices of `G` that are
-contained in some edge of matching `M`
+The subgraph `M` of `G` is a matching on `G` if every vertex only forms at most one edge of `M`.
 -/
-def matching.support (M : G.matching) : set V :=
-{v : V | ∃ x ∈ M.edges, v ∈ x}
+def is_matching : Prop := ∀ (v w w': V), M.adj v w → M.adj v w' → w = w'
 
 /--
-A perfect matching `M` on graph `G` is a matching such that
-  every vertex is contained in an edge of `M`.
+`M` is a perfect matching on `G` if it's a matching on `G` and if every vertex forms an edge of `M`.
 -/
-def matching.is_perfect (M : G.matching) : Prop :=
-M.support = set.univ
+def is_perfect : Prop := is_matching M ∧ M.support = set.univ
 
-lemma matching.is_perfect_iff (M : G.matching) :
-M.is_perfect ↔ ∀ (v : V), ∃ e ∈ M.edges, v ∈ e :=
-set.eq_univ_iff_forall
+lemma is_perfect_iff :
+  is_perfect M ↔ ∀ (v : V), ∃! (w : V), M.adj v w :=
+begin
+  simp only [is_matching, is_perfect, exists_unique],
+  split,
+  { intros h v,
+    cases h with h hu,
+    have hu' := set.mem_univ v,
+    rw [← hu, subgraph.mem_support] at hu',
+    cases hu' with w hvw,
+    use [w, hvw],
+    intros w' hvw',
+    specialize h v w w' hvw hvw',
+    simp only [h], },
+  { intro h,
+    split,
+    { intros v w w' hvw hvw',
+      specialize h v,
+      cases h with x h,
+      cases h with _ hy,
+      have hxw := hy w hvw,
+      have hxw' := hy w' hvw',
+      simp only [hxw, hxw'], },
+    { rw set.eq_univ_iff_forall,
+      intro v,
+      specialize h v,
+      cases h with w h,
+      cases h with hvw h,
+      rw subgraph.mem_support,
+      use w,
+      exact hvw, }, },
+end
```

Comment:
This lemma should be given as a corollary of another lemma, which is that M is a matching iff for all v in M.support, there exists a unique w in M.support such that M.adj v w.